lib/src/package_config_json.dart

```diff
@@ -0,0 +1,243 @@
+                      languageVersion);
+                }
+                Package package;
+                try {
+                  package = SimplePackage(
+                      name, rootUri, packageUriRootUri, languageVersion);
+                } on ArgumentError catch (e) {
+                  throw FormatException(e.message, e.invalidValue);
+                }
+                accumulator.add(package);
+                continue;
+              } else {
+                throw FormatException(
+                    root == null
+                        ? "Missing package root"
+                        : "Package root is not a string",
+                    root);
+              }
+            } else {
+              throw FormatException(
+                  name == null
+                      ? "Missing package name"
+                      : "Package name is not a string",
+                  name);
+            }
+          } else {
+            throw FormatException(
+                "Package entry is not a JSON object", packageEntry);
+          }
+        }
+        if (extraData != null) {
+          for (var key in map.keys) {
+            if (key != _configVersionKey && key != _packagesKey) {
+              extraData[key] = map[key];
+            }
+          }
+        }
+        try {
+          return SimplePackageConfig(version, accumulator);
+        } on ArgumentError catch (e) {
+          throw FormatException(e.message, e.invalidValue);
+        }
+      }
+      throw FormatException(
+          packageList == null
+              ? "Missing packages list"
+              : 'Invalid packages: Not a list',
+          packageList);
+    }
+    throw FormatException(
+        version == null
+            ? "Missing configVersion"
+            : "Invalid configVersion number",
+        version);
+  }
+  throw FormatException("Not a JSON object", map);
+}
+
+PackageConfig _parseDotPackagesConfig(Uint8List bytes, File file) {
+  return packagesfile.parse(bytes, Uri.file(file.path));
+}
+
+void writePackageConfigJson(PackageConfig config, Directory targetDirectory,
+    Map<String, dynamic> /*?*/ extraData) {
+  // Write .dart_tool/package_config.json first.
+  var file = File(
+      path.join(targetDirectory.path, ".dart_tool", "package_config.json"));
+  var baseUri = Uri.file(file.path);
+  var data = <String, dynamic>{
+    _configVersionKey: PackageConfig.maxVersion,
+    _packagesKey: [
+      for (var package in config.packages)
+        <String, dynamic>{
+          _nameKey: package.name,
+          _rootUriKey: relativizeUri(package.root, baseUri),
+          if (package.root != package.packageUriRoot)
+            _packageUriKey: relativizeUri(package.packageUriRoot, package.root),
+          if (package.languageVersion != null)
+            _languageVersionKey: package.languageVersion
+        }
+    ]
+  };
+
+  extraData?.forEach((key, value) {
+    if (key == _configVersionKey || key == _packagesKey) return; // Ignore.
```

Comment:
You could just check if `data.containsKey(key)` as well and that might be more future proof.

Reply:
Reasonable. It will also trigger if extraData somehow contains the same key twice, but behavior  in that case is not promised anyway.